File: sorting/selection_sort/selection_sort.py
def selection_sort(elements: list[int]) -> list[int]:
    for i in range(len(elements)):
        min_element_index: int = i

        for j in range(i, len(elements)):
            if elements[j] < elements[min_element_index]:
                min_element_index = j

        elements[i], elements[min_element_index] = elements[min_element_index], elements[i]

    return elements


def selection_sort_optimized(elements: list[int]) -> list[int]:
    length: int = len(elements)
    min_element_index: int = -1
    for i in range(length - 1):
        min_element_index = i

        for j in range(i + 1, length):
            if elements[j] < elements[min_element_index]:
                min_element_index = j

        elements[i], elements[min_element_index] = elements[min_element_index], elements[i]

    return elements


if __name__ == '__main__':
    input: list[int] = [8, 2, 2, 9, 1, 100, 0]
    print(input)
    print(f"{selection_sort(input[:]) = }")
    print(f"{selection_sort_optimized(input[:]) = }")

File: sorting/selection_sort/SelectionSort.java
class SelectionSort {
    public static void main(String[] args) {
        int[] input = { 1, 3, 5, 0, 8, 4, 2 };
        print(input);
        print(selectionSort(input.clone()), "selectionSort");
    }

    static int[] selectionSort(int[] elements) {
        int i, j, minElementIndex, temp;
        int length = elements.length;
        for (i = 0; i < length - 1; i++) {
            minElementIndex = i;
            for (j = i + 1; j < length; j++) {
                if (elements[j] < elements[minElementIndex]) {
                    minElementIndex = j;
                }
            }

            temp = elements[i];
            elements[i] = elements[minElementIndex];
            elements[minElementIndex] = temp;

        }
        return elements;
    }

    static void print(int[] array, String text) {
        System.out.print(text + " ");
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    static void print(int[] array) {
        for (int num : array) {
            System.out.print(num + " ");
        }
        System.out.println();

    }
}